### PR TITLE
Dynamo: Support min / max over iterables

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94350
* #94326

Expands support for built-in `min` and `max` calls beyond binary to iterables - simply reduce over the existing binary logic.
Adds support for:
* lists
* tuples
* list iterators
* vararg min / max - `min(2, 3, 4)`

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -581,7 +581,28 @@ def call_function(
             )
         return super().call_function(tx, args, kwargs)
 
-    def _call_min_max(self, tx, a, b):
+    def _call_min_max(self, tx, *args):
+        if len(args) == 1 and args[0].has_unpack_var_sequence(tx):
+            return self._call_min_max_iterable(tx, args[0])
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builtin.py:586]: Is this ever reached with the `len(items) ==  1` below? or am I misreading it?

File: torch/_dynamo/variables/builtin.py [hunk 2]
```diff
@@ -581,7 +581,28 @@ def call_function(
...
+        elif len(args) == 2:
+            return self._call_min_max_binary(tx, args[0], args[1])
+        else:
+            unimplemented(
+                f"unsupported min / max over {len(args)} args: {[str(a) for a in args]}"
+            )
+
+    def _call_min_max_iterable(self, tx, obj):
+        items = obj.unpack_var_sequence(tx)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builtin.py:595]: assert has_unpack_var_sequence - I know its checked above, but its good to protect the actual callsite :) it raises NotImplementedError if its not impelemented

File: torch/_dynamo/variables/builtin.py [hunk 3]
```diff
@@ -581,7 +581,28 @@ def call_function(
...
-    def _call_min_max(self, tx, a, b):
+    def _call_min_max(self, tx, *args):
+        if len(args) == 1 and args[0].has_unpack_var_sequence(tx):
+            return self._call_min_max_iterable(tx, args[0])
+        elif len(args) == 2:
+            return self._call_min_max_binary(tx, args[0], args[1])
+        else:
+            unimplemented(
+                f"unsupported min / max over {len(args)} args: {[str(a) for a in args]}"
+            )
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builtin.py:592]: This is a bit redundant with the else / fallthrough case below, doesn't matter too much tho

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/variables/builtin.py (hunk 3 above)
Comment (ezyang) [inline torch/_dynamo/variables/builtin.py:592]: varargs min is a thing and it doesn't seem like much trouble to support

```
>>> min(1,2,3)
1
```

File: torch/_dynamo/variables/builtin.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/variables/builtin.py:595]: idk, seems kind of pointless to me, if you've got it wrong unpack_var_sequence will blow up anyway, you aren't getting any safety from an extra assert here

File: torch/_dynamo/variables/builtin.py [hunk 4]
```diff
@@ -581,7 +581,28 @@ def call_function(
...
+
+    def _call_min_max_iterable(self, tx, obj):
+        items = obj.unpack_var_sequence(tx)
+        assert len(items) > 0
+        if len(items) == 1:
+            return items[0]
```
Comment (ezyang) [inline torch/_dynamo/variables/builtin.py:598]: You need to propagate guards from obj to this value!

**Review by ezyang — APPROVED** [review]
Insufficient options propagation

File: torch/_dynamo/variables/builtin.py (hunk 3 above)
Comment (jbschlosser) [reply torch/_dynamo/variables/builtin.py:592]: True, it's not even that much nicer of a message. I'll remove it

File: torch/_dynamo/variables/builtin.py (hunk 4 above)
Comment (jbschlosser) [reply torch/_dynamo/variables/builtin.py:598]: It's also done on the return value of the handler:
https://github.com/pytorch/pytorch/blob/fe0e28ab87ddda38a82efe74dd3eaf05e926f117/torch/_dynamo/variables/builtin.py#L563-L565

**Pull request closed without merging**